File: benchmark.jl
using Random
using Printf
using Dates

function randomBigInt(size)
    return rand(BigInt(0):BigInt(2)^size - 1)
end

function testRandomBigIntPerformance(w)
    for s in 6:(23)
        size=2^s
        e = 23 - floor(Int, round(log2(size))) + 1 + (w >= 2 ? 2 : 0)
        count = 2^max(e, 0)
        a = Vector{BigInt}(undef, count)
        b = Vector{BigInt}(undef, count)

        for i in 1:count
            a[i] = randomBigInt(w == 3 || w == 4 ? size * 2 : size)
            b[i] = randomBigInt(w == 3 || w == 4 ? size - 2 : size)
            #while (gcd(a[i], b[i]) != 1)
            #   b[i] += 1
            #end
        end

        start = now()
        sum = 0
        g = BigInt(0)
        for i in 1:count
            if (w == 0)
              g = gcd(a[i], b[i])
            end
            if (w == 1)
              #g, u, v = gcdx(a[i], b[i])
              g = invmod(a[i], b[i])
            end
            if (w == 2)
              g = a[i] * b[i]
            end
            if (w == 3)
              g = a[i] ÷ b[i]
            end
            if (w == 4)
              g = a[i] % b[i]
            end
            #if (w == 5)
            #  g = invmod(a[i], b[i])
            #end
            sum += Int(g & 0xFFFF)
        end
        elapsed = (now() - start) / Millisecond(1)
        @printf("%d %.5f0ms %d\n", size, elapsed / count, sum)
    end
end

testRandomBigIntPerformance(0)
testRandomBigIntPerformance(1)
testRandomBigIntPerformance(2)
testRandomBigIntPerformance(3)
testRandomBigIntPerformance(4)
#testRandomBigIntPerformance(5)

File: benchmark.py
import random
import gmpy2
from gmpy2 import gcd, invert, mpz
from datetime import datetime

def random_bigint(size):
    return mpz(random.getrandbits(size))

def test_random_bigint_performance(w):
    for s in range(6, 24):
        size = 2**s
        e = 23 - round(size.bit_length() - 1) + 1 + (2 if w >= 2 else 0)
        count = 2**max(e, 0)
        a = [None] * count
        b = [None] * count

        for i in range(count):
            a[i] = random_bigint(size * 2 if w in [3, 4] else size)
            b[i] = random_bigint(size - 2 if w in [3, 4] else size)

        start = datetime.now()
        total_sum = 0
        g = mpz(0)

        for i in range(count):
            if w == 0:
                g = gcd(a[i], b[i])
            elif w == 1:
                try:
                    g = invert(a[i], b[i])
                except ZeroDivisionError:
                    g = mpz(0)
            elif w == 2:
                g = a[i] * b[i]
            elif w == 3:
                g = a[i] // b[i]
            elif w == 4:
                g = a[i] % b[i]
            total_sum += int(g & 0xFFFF)

        elapsed = (datetime.now() - start).total_seconds() * 1000  # in milliseconds
        print(f"{size} {elapsed / count:.5f}ms {total_sum}")

# Test the performance for different operations
test_random_bigint_performance(0)
test_random_bigint_performance(1)
test_random_bigint_performance(2)
test_random_bigint_performance(3)
test_random_bigint_performance(4)